---
title: "Seurat Analysis"
output: html_document
---

```{r setup, include=FALSE}
Sys.setenv(RSTUDIO_PANDOC="/usr/lib/rstudio-server/bin/pandoc")
knitr::opts_chunk$set(echo = TRUE)
```

## **Read in packages**

```{r, echo=FALSE}
library(Matrix)
library(dplyr)
library(Seurat)
library(patchwork)
library(rmarkdown)
library(devtools)
library(SingleR)
library(scCATCH)
library(cowplot)
library(ggplot2)
```


## **Seurat Visualizations and Analyses**


```{r}
path <- "/project/Dolatshahi_Lab/MSDS/Clean_Final/cellranger_pbmc/filtered_feature_bc_matrix/"
barcode.path <- paste0(path, "barcodes.tsv.gz")
features.path <- paste0(path, "features.tsv.gz")
matrix.path <- paste0(path, "matrix.mtx.gz")
mat <- readMM(file = matrix.path)
feature.names = read.delim(features.path,
                           header = FALSE,
                           stringsAsFactors = FALSE)
barcode.names = read.delim(barcode.path,
                           header = FALSE,
                           stringsAsFactors = FALSE)
colnames(mat) = barcode.names$V1
rownames(mat) = feature.names$V2

testseurat <- CreateSeuratObject(counts = mat , min.cells = 3, min.features = 200, project = "10X_PBMC")
 
testseurat[["percent.mt"]] <- PercentageFeatureSet(testseurat, pattern = "^MT-")
```

```{r}
VlnPlot(testseurat, features = c("nFeature_RNA", "nCount_RNA", "percent.mt"), ncol = 3)
```

```{r}
plot1 <- FeatureScatter(testseurat, feature1 = "nCount_RNA", feature2 = "percent.mt")
plot2 <- FeatureScatter(testseurat, feature1 = "nCount_RNA", feature2 = "nFeature_RNA")
plot1 + plot2
```

```{r}
x <- min(25, 2*sd(testseurat[["percent.mt"]][[1]])+mean(testseurat[["percent.mt"]][[1]]))
paste0("Filtering percent.mt: ", x)
newseurat <- subset(testseurat, subset = nFeature_RNA > 200 & nFeature_RNA < 4500 & percent.mt < x)
```


```{r}
normalized_object <- NormalizeData(newseurat, normalization.method = "LogNormalize", scale.factor = 10000)
plot1 <- FeatureScatter(normalized_object, feature1 = "nCount_RNA", feature2 = "nFeature_RNA")
plot1
```

```{r}
# how to choose top variable features. Picking genes that differ the most between cells in terms of expression
newseurat <- FindVariableFeatures(normalized_object, selection.method = "vst", nfeatures = 1000)
top10 <- head(VariableFeatures(newseurat), 10)
plot1 <- VariableFeaturePlot(newseurat)
plot2 <- LabelPoints(plot = plot1, points = top10, repel = TRUE)
plot1
```


```{r}
plot2
```

```{r}
all.genes <- rownames(newseurat)
newseurat <- ScaleData(newseurat, features = all.genes)
newseurat <- RunPCA(newseurat, features = VariableFeatures(object = newseurat))

print(newseurat[['pca']], dims=1:5, nfeatures=5)
VizDimLoadings(newseurat, dims = 1:2, reduction = "pca")

```



```{r}
DimPlot(newseurat, reduction = "pca")

```

```{r}
DimHeatmap(newseurat, dims = 1, cells = 500, balanced = TRUE)
```

```{r}
DimHeatmap(newseurat, dims = 1:15, cells = 500, balanced = TRUE)
```

```{r}
newseurat <- JackStraw(newseurat, num.replicate = 100)
newseurat <- ScoreJackStraw(newseurat, dims = 1:20)
JackStrawPlot(newseurat, dims = 1:20)
```

```{r}
ElbowPlot(newseurat)
```

```{r}
newseurat <- FindNeighbors(newseurat, dims = 1:10)
```


```{r}
newseurat <- FindClusters(newseurat, resolution = .5)
head(Idents(newseurat), 5)
```

## **PanglaoDB Method**

### <font color = "red">**NOTE: This method uses data that has first been processed with cellranger, rather than the preprocessed version provided by the Seurat tutorial.**</font>

```{r}
library(data.table)
library(magrittr)

## extract meta data
md <- newseurat@meta.data %>% as.data.table
```


```{r}
newseurat <- RunUMAP(newseurat, dims = 1:10)
```


```{r}
# Identify gene markers
all_markers<- FindAllMarkers(object = newseurat, only.pos = TRUE, min.pct = 0.25, logfc.threshold = 0.25)
```
Select top 3 markers, ordered by the meteric avg_log2FC

```{r}
markers <- all_markers %>%
    group_by(cluster) %>%
    slice_max(n = 3, order_by = avg_log2FC) %>%
    select(cluster, gene)
```

Read in data from PanglaoDB: 

```{r, warning=FALSE}
cell_types = read.csv('immune_marker.csv', header = FALSE)
cell_types[is.na(cell_types)] <-0
```


```{r, warning=FALSE}
cell_types = cell_types %>% select(c('V2','V3', 'V11','V13'))
colnames(cell_types) <- c('Gene marker','Cell Type','Sensitivity','Specificity')
cell_types$Sensitivity[cell_types$Sensitivity==0] <- mean(cell_types$Sensitivity)-0.1*mean(cell_types$Sensitivity) # add penalty to missing values
cell_types$Specificity[cell_types$Specificity==0] <- mean(cell_types$Specificity)+0.1*mean(cell_types$Specificity)

```


```{r}
markers <- left_join(markers, cell_types, by = c("gene" = "Gene marker"))
head(markers)
```

```{r}
markers$`Cell Type`[is.na(markers$`Cell Type`)]<-'Unknown'
markers$Sensitivity[is.na(markers$Sensitivity)]<-0
markers$Specificity[is.na(markers$Specificity)]<-0
```


```{r}
markers$range = markers$Sensitivity-markers$Specificity
head(markers)
```

```{r}
cluster_info <- markers %>% slice_max(n=1, order_by = range, with_ties = FALSE) %>% select(c('cluster', 'Cell Type'))
cluster_info
```

```{r}
new_cluster_ids <- cluster_info$`Cell Type`
```

```{r}
names(new_cluster_ids) <- levels(newseurat)
newseurat <- RenameIdents(newseurat, new_cluster_ids)
panglaodb_plot = DimPlot(newseurat, reduction = "umap", repel = T, label = TRUE, pt.size = 0.5)
no_legend_panglaodb_plot = DimPlot(newseurat, reduction = "umap", repel = T, label = TRUE, pt.size = 0.5) + NoLegend()
panglaodb_plot
```

```{r}
cell_count <- md %>% group_by(seurat_clusters) %>% summarise(count=n()) %>% mutate(percent = count/sum(count)*100) %>% arrange(desc(percent))
```

```{r}
joined_table = left_join(cell_count, cluster_info, by = c("seurat_clusters" = "cluster"))
joined_table
```

Individual Cell and its respective cell type: 

```{r}
barcode = barcode.names
colnames(barcode) <- "Cell"
barcode$count = seq(1, length(barcode$Cell))
```

```{r}
md$count = seq(1, length(md$seurat_clusters))
new_md = inner_join(barcode, md, by=c("count"="count"))
```

```{r}
cellwithtype = inner_join(joined_table, new_md, by=c("seurat_clusters"="seurat_clusters"))
final_df = subset(cellwithtype, select=c("Cell", "Cell Type"))
colnames(final_df) <- c('Cell', 'Cell_Types')
```

```{r}
write.csv(final_df, file = 'PanglaoDB_Individual_Cell_with_its_Cell_Type.csv', row.names = FALSE)
```


## **SingleR Method**


### **Use package `celldex` to get built-in references to annotate cells... but mostly from bulk RNA-seq or microarray data

**Source**: [Using SingleR to annotate single-cell RNA-seq data](https://bioconductor.org/packages/devel/bioc/vignettes/SingleR/inst/doc/SingleR.html)


For first time installation: 
install.packages("remotes")
remotes::install_github("LTLA/celldex")

```{r}
library(celldex)
```

```{r}
# Human Primary Cell Atlas (Mabbott et al. 2013) with a matrix of log-expression values with sample-level labels
hpca.se = HumanPrimaryCellAtlasData()
```

```{r}
results = SingleR(method = 'single', test = as.SingleCellExperiment(newseurat), ref = hpca.se, labels = hpca.se$label.main)
```


```{r}
unique(results$pruned.labels)
```


```{r}
newseurat$SingleR.pruned.calls = results$pruned.labels
newseurat$SingleR.calls = results$labels
```

#### **Source**: Ch.8: Single cell RNA-seq analysis using Seurat $\rightarrow$ 8.5: Cell type annotation using SingleR: https://www.singlecellcourse.org/single-cell-rna-seq-analysis-using-seurat.html

```{r}
newseurat = SetIdent(newseurat, value = "SingleR.calls")
singler_plot = DimPlot(newseurat, reduction = "umap", label = TRUE, repel = T, label.size = 3)
no_legend_singler_plot = DimPlot(newseurat, reduction = "umap", repel = T, label.size = 3, pt.size = 0.5) + NoLegend()
singler_plot
```

```{r}

cells = names(newseurat$SingleR.pruned.calls)

labels = unname(newseurat$SingleR.pruned.calls)

cell_labels = tibble(cell = cells, Cell_Types = labels)

cell_labels
```

```{r}
cell_labels %>% group_by(Cell_Types) %>% summarise(count=n()) %>% mutate(percent = count/sum(count)*100) %>% arrange(desc(percent))
```



## **scCATCH Method**

scCATCH Github Repository: https://github.com/ZJUFanLab/scCATCH


```{r}
newseurat <- subset(testseurat, subset = nFeature_RNA > 200 & nFeature_RNA < 4500 & percent.mt < x)
```

```{r}
normalized_object <- NormalizeData(newseurat, normalization.method = "LogNormalize", scale.factor = 10000)
```

```{r}
# how to choose top variable features. Picking genes that differ the most between cells in terms of expression
newseurat <- FindVariableFeatures(normalized_object, selection.method = "vst", nfeatures = 1000)
```


```{r, message=FALSE}
all.genes <- rownames(newseurat)
newseurat <- ScaleData(newseurat, features = all.genes)
newseurat <- RunPCA(newseurat, features = VariableFeatures(object = newseurat))
```


```{r}
newseurat <- FindNeighbors(newseurat, dims = 1:10)
newseurat <- FindClusters(newseurat, resolution = .5)
```

```{r, message=FALSE}
newseurat <- RunUMAP(newseurat, dims = 1:10)
```



```{r}
# create the first input for scCATCH
data = GetAssayData(object = normalized_object)

# create the second input for scCATCH
clusters = Idents(newseurat)
cluster_info = as.character((unname(clusters)))
```

Available species and tissues: https://github.com/ZJUFanLab/scCATCH/wiki/human_tissues


```{r, message=FALSE, warning=FALSE}
# create scCATCH object: 
obj = createscCATCH(data = data, cluster = cluster_info)

# Running scCATCH
cellmatch_new <- cellmatch[cellmatch$species == "Human" & cellmatch$tissue %in% c("Blood", "Peripheral blood"),]
clu_markers <- findmarkergene(object = obj, if_use_custom_marker = TRUE, marker = cellmatch_new)
                               
clu_ann <- findcelltype(object = clu_markers)
```

```{r}
clu_ann@celltype$cell_type
```


```{r}
tmp1 <- data.frame(cluster = levels(Idents(newseurat)))
tmp <- merge(tmp1, clu_ann@celltype, by = 'cluster', all = T)
tmp$cell_type[which(is.na(tmp$cell_type))] <- "Unclassified"
```

```{r}
new.cluster.ids <- tmp$cell_type
names(new.cluster.ids) <- levels(newseurat)
newseurat <- RenameIdents(newseurat, new.cluster.ids)
sccatch_plot = DimPlot(newseurat, reduction = "umap", repel = TRUE, label = TRUE, pt.size = 0.5)
no_legend_sccatch_plot = DimPlot(newseurat, reduction = "umap", repel = T, label.size = 3, pt.size = 0.5) + NoLegend()
sccatch_plot
```


```{r}
catch_df = as.data.frame(clu_ann@marker$celltype)
colnames(catch_df) <- "Cell_Types"
```


Cell Type Composition by Count and Percentage: 

Individual cell and its respective cell type:

```{r}
cell_df = as.data.frame(clu_ann@meta)
colnames(cell_df) <- c('Cell', 'cluster')
```

```{r}
cellwithtype = inner_join(cell_df, tmp, by=c("cluster"="cluster"))
scCatch_final_df = subset(cellwithtype, select=c("Cell", "cell_type"))
colnames(scCatch_final_df) <- c('Cell', 'Cell_Types')
```

```{r}
write.csv(scCatch_final_df, file = 'scCATCH_Individual_Cell_with_its_Cell_Type.csv', row.names = FALSE)
```

```{r}
scCatch_final_df %>% group_by(Cell_Types) %>% summarise(count=n()) %>% mutate(percent = count/sum(count)*100) %>% arrange(desc(percent))
```



## **Plot with UMAP for All Methods**


```{r}
par(mfrow=c(2,2))
panglaodb_plot
sccatch_plot
singler_plot
```



```{r}
# ggsave(filename = "pbmc.jpg", height = 20, width = 29, plot = full_plot, quality = 100)
```

```{r}
# ggsave(filename = "pbmc_top.jpg", height = 20, width = 50, plot = top_three_plots, quality = 100)
# ggsave(filename = "pbmc_bottom.jpg", height = 20, width = 29, plot = bottom_two_plots, quality = 100)
```

